ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Iterator.hs

```diff
@@ -0,0 +1,301 @@
+            ppStream expectedStream
+  where
+    (_trace, actual) = runIterator setup from to
+    failure msg = counterexample msg False
+
+    ppStream :: [Either (HeaderHash TestBlock) TestBlock] -> String
+    ppStream = intercalate " :> " . map ppEBBOrBlock
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+-- | The initial contents of the ImmutableDB and the VolatileDB.
+--
+-- Note that the iterator implementation does not rely on the current
+-- in-memory chain.
+data TestSetup = TestSetup
+  { immutable :: Chain TestBlock
+  , volatile  :: [TestBlock]
+  }
+
+mkBlk :: [Word64] -> TestBlock
+mkBlk h = TestBlock
+    { tbHash  = mkTestHash h
+    , tbSlot  = SlotNo $ fromIntegral $ 2 * length h
+    , tbValid = True
+    }
+
+-- | Human-friendly string description of the 'TestSetup' that can be used
+-- when printing a failing test.
+testSetupInfo :: TestSetup -> String
+testSetupInfo TestSetup { immutable, volatile } = mconcat
+    [ "Immutable: "
+    , intercalate " :> " (map ppBlock (Chain.toOldestFirst immutable))
+    , "\n"
+    , "Volatile:  "
+    , intercalate ", " (map ppBlock volatile)
+    ]
+
+ppEBBOrBlock :: Either (HeaderHash TestBlock) TestBlock -> String
+ppEBBOrBlock (Left  ebbHash) = "EBB " <> condense ebbHash
+ppEBBOrBlock (Right blk)     = ppBlock blk
+
+ppBlock :: TestBlock -> String
+ppBlock = condense . blockHash
+
+{-------------------------------------------------------------------------------
+  Running an iterator test
+-------------------------------------------------------------------------------}
+
+type IterRes = Either (UnknownRange TestBlock)
+                      [Either (HeaderHash TestBlock) TestBlock]
+                      -- Left:  EBB hash
+                      -- Right: regular block
+
+-- | Open an iterator with the given bounds on the given 'TestSetup'. Return a
+-- trace of the 'TraceIteratorEvent's produced and the result of the iterator
+-- itself.
+runIterator
+  :: TestSetup
+  -> StreamFrom TestBlock
+  -> StreamTo   TestBlock
+  -> ([TraceIteratorEvent TestBlock], IterRes)
+runIterator setup from to = runSimOrThrow $ do
+    (tracer, getTrace) <- recordTrace
+    itEnv <- initIteratorEnv setup tracer
+    res <- runExceptT $ do
+      it <- ExceptT $ newIterator itEnv ($ itEnv) from to
+      lift $ consume it
+    trace <- getTrace
+    return (trace, res)
+  where
+    consume :: Monad m
+            => Iterator m TestBlock
+            -> m [Either TestHash TestBlock]
+    consume it = iteratorNext it >>= \case
+      IteratorResult blk -> (Right blk :) <$> consume it
+      IteratorBlockGCed hash -> do
+        iteratorClose it
+        return [Left hash]
+      IteratorExhausted -> do
+        iteratorClose it
+        return []
+
+recordTrace :: MonadSTM m => m (Tracer m ev, m [ev])
```

Comment:
I noticed last week that this function wasn't in `Control.Tracer`. Does it belong there?

Reply:
I also have an [implementation](https://github.com/input-output-hk/ouroboros-network/blob/bfbb919c063678a35e066e58b90c25e972458a63/ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs#L1011) of this function using `IORef` (to be used with real `IO`, of course). That one could become part of `Control.Tracer`.

This function, however, would require a dependency on `io-sim-classes` (thanks to `MonadSTM`), and I don't know whether [they](https://github.com/input-output-hk/iohk-monitoring-framework/blob/master/contra-tracer/src/Control/Tracer.lhs) would like that :slightly_smiling_face:.

We could introduce `Test.Util.Tracer` to house these two.

Reply:
Ah, right, `MonadSTM`, not `STM`. `Test.Util.Tracer` seems right to me, though I do have a tendency to jump-the-gun wrt DRY.

Reply:
Done